src/services/SearchService.ts

```diff
@@ -1,7 +1,243 @@
+  SpaceRuleRepository
+} from '../repositories/RuleRepository';
+import { DayOfWeekLOSRule, NoticeRequiredRule } from '../proto/lpms';
+import { FormattedDate, RulesItemKey } from './DBService';
+import { SpaceAvailabilityRepository } from '../repositories/SpaceAvailabilityRepository';
+import { SpaceStubRepository } from '../repositories/SpaceStubRepository';
+import ApiError from '../exceptions/ApiError';
 
 export class SearchService {
-  public search = async (facilityId: string, ask: Ask): Promise<string[]> => {
-    throw new Error('Not Implemented Yet');
+  protected facilityRuleRepository: FacilityRuleRepository;
+  protected facilityId: string;
+  protected ask: Ask;
+
+  constructor(facilityId: string, ask: Ask) {
+    this.facilityRuleRepository = new FacilityRuleRepository(facilityId);
+    this.facilityId = facilityId;
+    this.ask = ask;
+  }
+
+  public search = async (): Promise<string[]> => {
+    const spacesIds = await facilityRepository.getFacilityKey(
+      this.facilityId,
+      'spaces'
+    );
+
+    if (Array.isArray(spacesIds)) {
+      const set = new Set();
+
+      for (const v of spacesIds) {
+        const space = await facilityRepository.getItemKey(
+          this.facilityId,
+          'spaces',
+          v,
+          'metadata'
+        );
+        set.add({ space, id: v });
+      }
+
+      return this.findSpaces(Array.from(set));
+    }
+
+    return [];
   };
+
+  private async findSpaces(spaces): Promise<string[]> {
+    const needed = new Set<string>();
+
+    const dates = this.getAskDates();
+    if (!dates.length) return [];
+
+    for (const i of spaces) {
+      const space = i.space as Space;
+
+      //check rules
+      if (!(await this.checkRules(dates, i.id))) {
+        continue;
+      }
+
+      //check space capacity
+      if (!this.checkSuitableQuantity(space)) {
+        continue;
+      }
+
+      //check dates is available
+      if (await this.checkAvailableDates(i.id, dates)) {
+        needed.add(i.id);
+      }
+    }
+
+    return Array.from(needed);
+  }
+
+  private getAskDates(): DateTime[] {
+    if (!this.ask.checkIn || !this.ask.checkOut) {
+      throw ApiError.BadRequest('invalid dates in ask');
+    }
+
+    let from = DateTime.fromObject(this.ask.checkIn);
+    const to = DateTime.fromObject(this.ask.checkOut);
+    const dates: DateTime[] = [];
+
+    while (from <= to) {
+      dates.push(from);
+      from = from.plus({ days: 1 });
+    }
+
+    return dates;
+  }
+
+  private async checkRules(
+    dates: DateTime[],
+    spaceId: string
+  ): Promise<boolean> {
+    const spaceRuleRepository = new SpaceRuleRepository(
+      this.facilityId,
+      spaceId
+    );
+
+    const noticeRequirementRule = (await this.getRule(
+      'notice_required',
+      spaceRuleRepository
+    )) as NoticeRequiredRule;
+
+    const firstDay = dates[0];
+
+    const today = DateTime.now();
+    const interval = Interval.fromDateTimes(today, firstDay).toDuration([
```

Comment:
This should also take into account the standard check-in time for the facility provider which is in the metadata, so specifically this should be thought of as like '2 days = 48 hrs before'. Timezones should also be taken into account here (need to double-check the metadata provides for timezone).